man/tpm2_zgen2phase.1.md
% tpm2_zgen2phase(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_zgen2phase**(1) - Command to enable the TPM to combine data from the
other party with the ephemeral key generated in the first phase of two-phase
key exchange protocols.


# SYNOPSIS

**tpm2_zgen2phase** [*OPTIONS*]

# DESCRIPTION

**tpm2_zgen2phase**(1) - Command to enable the TPM to combine data from the
other party with the ephemeral key generated in the first phase of two-phase
key exchange protocols.

# OPTIONS

  * **-c**, **\--key-context**=_FILE_:

    Context object pointing to ECC key.
    Either a file or a handle number. See section "Context Object Format".

  * **-p**, **\--key-auth**=_AUTH_:

    The authorization value for the ECC key object.

  * **-s**, **\--scheme**=_ALGORITHM_:

    The key exchange scheme. Optional. Valid options are ecdh or sm2.

  * **-t**, **\--counter**=_NATURALNUMBER_:

    The commit count to determine the key index to use.

  * **\--static-public**=_FILE_:

    The static public key input of the other party.

  * **\--ephemeral-public**=_FILE_:

    The ephemeral public key input of the other party.

  * **\--output-Z1**=_FILE_

    Specify file path to save the calculated ecdh secret Z1 point.

  * **\--output-Z2**=_FILE_

    Specify file path to save the calculated ecdh secret Z2 point.

## References

[algorithm specifiers](common/alg.md) details the options for specifying
cryptographic algorithms _ALGORITHM_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.

# EXAMPLES

```bash
tpm2_createprimary -C o -c prim.ctx -Q

tpm2_create -C prim.ctx -c key.ctx -u key.pub -r key.priv -G ecc256:ecdh -Q

tpm2_ecephemeral -u ecc.q -t ecc.ctr ecc256

tpm2_ecdhkeygen -u ecdh.pub -o ecdh.priv -c key.ctx

tpm2_zgen2phase -c key.ctx --static-public ecdh.pub --ephemeral-public ecc.q \
-t 0 --output-Z1 z1.bin --output-Z2 z2.bin
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_zgen2phase.c
/* SPDX-License-Identifier: BSD-3-Clause */
#include "files.h"
#include "log.h"
#include "object.h"
#include "tpm2.h"
#include "tpm2_tool.h"
#include "tpm2_auth_util.h"
#include "tpm2_alg_util.h"
#include "tpm2_options.h"

typedef struct tpm_ecdhzgen_ctx tpm_ecdhzgen_ctx;
struct tpm_ecdhzgen_ctx {

    struct {
        const char *ctx_path;
        const char *auth_str;
        tpm2_loaded_object object;
    } ecc_key;

    const char *output_z1_path;
    const char *output_z2_path;
    TPM2B_ECC_POINT *Z1;
    TPM2B_ECC_POINT *Z2;

    const char *static_public_path;
    const char *ephemeral_public_path;
    TPM2B_ECC_POINT Q1;
    TPM2B_ECC_POINT Q2;

    UINT16 commit_counter;
    TPMI_ECC_KEY_EXCHANGE keyexchange_scheme;
};

static tpm_ecdhzgen_ctx ctx = {
    .keyexchange_scheme = TPM2_ALG_ECDH,
};

static bool on_option(char key, char *value) {

    bool result = true;
    switch (key) {
        case 'c':
            ctx.ecc_key.ctx_path = value;
            break;
        case 'p':
            ctx.ecc_key.auth_str = value;
            break;
        case 's':
            ctx.keyexchange_scheme = tpm2_alg_util_from_optarg(value,
                tpm2_alg_util_flags_sig);
            break;
        case 't':
            result = tpm2_util_string_to_uint16(value, &ctx.commit_counter);
            if (!result) {
                LOG_ERR("Could not convert commit counter to number, got: \"%s\"",
                        value);
                return false;
            }
            break;
        case 0:
            ctx.static_public_path = value;
            break;
        case 1:
            ctx.ephemeral_public_path = value;
            break;
        case 2:
            ctx.output_z1_path = value;
            break;
        case 3:
            ctx.output_z2_path = value;
            break;
    };

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    static struct option topts[] = {
      { "key-context",      required_argument, NULL, 'c' },
      { "key-auth",         required_argument, NULL, 'p' },
      { "scheme",           required_argument, NULL, 's' },
      { "counter",          required_argument, NULL, 't' },
      { "static-public",    required_argument, NULL,  0  },
      { "ephemeral-public", required_argument, NULL,  1  },
      { "output-Z1",        required_argument, NULL,  2  },
      { "output-Z2",        required_argument, NULL,  3  },
    };

    *opts = tpm2_options_new("c:p:s:t:", ARRAY_LEN(topts), topts,
            on_option, NULL, 0);

    return *opts != NULL;
}

static tool_rc check_options(void) {

    if (ctx.keyexchange_scheme != TPM2_ALG_ECDH &&
        ctx.keyexchange_scheme != TPM2_ALG_ECMQV &&
        ctx.keyexchange_scheme != TPM2_ALG_SM2) {
            LOG_ERR("Unknown signing scheme");
            return tool_rc_general_error;
    }

    if (!ctx.ecc_key.ctx_path) {
        LOG_ERR("Specify an ecc key handle for context");
        return tool_rc_option_error;
    }

    if (!ctx.static_public_path) {
        LOG_ERR("Specify path to read the static public data from.");
        return tool_rc_option_error;
    }

    if (!ctx.ephemeral_public_path) {
        LOG_ERR("Specify path to read the ephemeral public data from.");
        return tool_rc_option_error;
    }

    if (!ctx.output_z1_path) {
        LOG_ERR("Specify path to save the Z1 data.");
        return tool_rc_option_error;
    }

    if (!ctx.output_z2_path) {
        LOG_ERR("Specify path to save the Z2 data.");
        return tool_rc_option_error;
    }

    return tool_rc_success;
}

static tool_rc process_inputs(ESYS_CONTEXT *ectx) {

    tool_rc rc = tpm2_util_object_load_auth(ectx, ctx.ecc_key.ctx_path,
        ctx.ecc_key.auth_str, &ctx.ecc_key.object, false,
        TPM2_HANDLES_FLAGS_TRANSIENT|TPM2_HANDLES_FLAGS_PERSISTENT);
    if (rc != tool_rc_success) {
        return rc;
    }

    bool result = true;
    result = files_load_ecc_point(ctx.static_public_path, &ctx.Q1);
    if (!result) {
        LOG_ERR("Failed to load static public input ECC point Q1");
        return tool_rc_general_error;
    }

    result = files_load_ecc_point(ctx.ephemeral_public_path, &ctx.Q2);
    if (!result) {
        LOG_ERR("Failed to load static public input ECC point Q2");
        return tool_rc_general_error;
    }

    return tool_rc_success;
}

static tool_rc process_outputs(void) {

    bool result = files_save_ecc_point(ctx.Z1, ctx.output_z1_path);
    if (!result) {
        LOG_ERR("Failed to write out the ECC point Z1");
        return tool_rc_general_error;
    }

    result = files_save_ecc_point(ctx.Z2, ctx.output_z2_path);
    if (!result) {
        LOG_ERR("Failed to write out the ECC point Z2");
        return tool_rc_general_error;
    }

    return tool_rc_success;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    UNUSED(flags);

    // Check input options and arguments
    tool_rc rc = check_options();
    if (rc != tool_rc_success) {
        return rc;
    }

    // Process inputs
    rc = process_inputs(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    // ESAPI call
    rc = tpm2_zgen2phase(ectx, &ctx.ecc_key.object, &ctx.Q1, &ctx.Q2, &ctx.Z1,
    &ctx.Z2, ctx.keyexchange_scheme, ctx.commit_counter);
    if (rc != tool_rc_success) {
        return rc;
    }

    // Process ouputs
    rc = process_outputs();

    return rc;
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("zgen2phase", tpm2_tool_onstart, tpm2_tool_onrun, NULL, NULL)
